jetbrains-core/src/software/aws/toolkits/jetbrains/core/stack/TreeView.kt

```diff
@@ -0,0 +1,152 @@
+package software.aws.toolkits.jetbrains.core.stack
+
+import com.intellij.ide.util.treeView.NodeDescriptor
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.project.Project
+import com.intellij.ui.components.JBScrollPane
+import com.intellij.ui.treeStructure.Tree
+import one.util.streamex.StreamEx
+import software.amazon.awssdk.services.cloudformation.model.StackResource
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import javax.swing.Icon
+import javax.swing.JComponent
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+import javax.swing.tree.MutableTreeNode
+
+/**
+ *
+ * Stack status tree
+ * @property component with tree content
+ */
+internal interface TreeView {
+    /**
+     * Update list of resources
+     */
+    fun fillResources(resources: Collection<StackResource>)
+
+    /**
+     * Sets whole stack status.
+     */
+    fun setStackStatus(status: StackStatus)
+
+    val component: JComponent
+}
+
+internal class TreeViewImpl(private val project: Project, stackName: String) : TreeView, Disposable, ViewWithIcons {
+
+    private val tree: Tree
+    override val component: JComponent
+
+    private val rootNode: DefaultMutableTreeNode
+        get() = tree.model.root as DefaultMutableTreeNode
+
+    private val rootDescriptor: StackNodeDescriptor
+        get() = rootNode.userObject as StackNodeDescriptor
+
+    private var model: DefaultTreeModel
+
+    /**
+     * Remove outdated nodes, add new nodes, update status
+     */
+    private fun updateResourceList(resources: Collection<StackResource>) {
+
+        val resourcesByName = resources.map { it.logicalResourceId() to it }.toMap()
+        val existingResources = HashSet<String>()
+        val nodesToDelete = ArrayList<MutableTreeNode>()
+
+        for (rawNode in rootNode.children()) {
+            val node = rawNode as DefaultMutableTreeNode
+            val descriptor = (node.userObject as StackNodeDescriptor)
+            val name = descriptor.element
+            val resource = resourcesByName[name]
+            if (resource != null) {
+                // Set status
+                val status = resource.resourceStatus()
+                descriptor.setStatusAndType(status.type, status.name)
+                descriptor.update()
+                existingResources.add(name)
+            } else {
+                nodesToDelete.add(node)
+            }
+        }
+        nodesToDelete.forEach { rootNode.remove(it) }
+        for (nameAndResource in resourcesByName) {
+            val name = nameAndResource.key
+            if (name in existingResources) {
+                continue
+            }
+            val resource = nameAndResource.value
+
+            val status = resource.resourceStatus()
+            val newDescriptor = StackNodeDescriptor(project, name, status.type, status.name, rootDescriptor)
+            newDescriptor.isWasDeclaredAlwaysLeaf = true
```

Comment:
My initial thoughs was that it is used for optimization to tell our ``Tree`` code that it can't have children (always leaf), but I discussed it with colleagues and it seems that no optimzation takes place here. So, I remove it because it is useless. Thanks for paying attention on it.